joblib/parallel.py

```diff
@@ -585,8 +598,18 @@ def _dispatch(self, batch):
 
         dispatch_timestamp = time.time()
         cb = BatchCompletionCallBack(dispatch_timestamp, len(batch), self)
-        job = self._backend.apply_async(batch, callback=cb)
-        self._jobs.append(job)
+        with self._lock:
+            job_idx = len(self._jobs)
+            job = self._backend.apply_async(batch, callback=cb)
+            # If the callback is already done and we are not using a
+            # SequentialBackend, we need to reorder the jobs to get the correct
+            # order in the result. This can happen when a job is so quick
+            # that its callback is called before we get here.
```

Comment:
The existing tests with fast functions (e.g. compute square) were able to highlight this potential problem with the concurrent future callback system.

The `multiprocessing.Pool.apply_async` callback system did not have this issue because the callback is always called in the callback thread and never in the main thread.